cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,484 @@
+		copy(nodeInfoList, baseList)
+	}
+
+	for _, ani := range data.addedNodeInfoMap {
+		nodeInfoList = append(nodeInfoList, ani)
+	}
+
+	return nodeInfoList
+}
+
+// Convenience method to avoid writing loop for adding nodes.
+func (data *internalDeltaSnapshotData) addNodes(nodes []*apiv1.Node) error {
+	for _, node := range nodes {
+		if err := data.addNode(node); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) addNode(node *apiv1.Node) error {
+	nodeInfo := schedulernodeinfo.NewNodeInfo()
+	if err := nodeInfo.SetNode(node); err != nil {
+		return fmt.Errorf("cannot set node in NodeInfo: %v", err)
+	}
+	return data.addNodeInfo(nodeInfo)
+}
+
+func (data *internalDeltaSnapshotData) addNodeInfo(nodeInfo *schedulernodeinfo.NodeInfo) error {
+	if _, found := data.getNodeInfo(nodeInfo.Node().Name); found {
+		return fmt.Errorf("node %s already in snapshot", nodeInfo.Node().Name)
+	}
+	data.addedNodeInfoMap[nodeInfo.Node().Name] = nodeInfo
+	if data.nodeInfoList != nil {
+		data.nodeInfoList = append(data.nodeInfoList, nodeInfo)
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) clearCaches() {
+	data.nodeInfoList = nil
+	data.clearPodCaches()
+}
+
+func (data *internalDeltaSnapshotData) clearPodCaches() {
+	data.podList = nil
+	data.havePodsWithAffinity = nil
+}
+
+func (data *internalDeltaSnapshotData) updateNode(node *schedulernodeinfo.NodeInfo) error {
+	if _, found := data.addedNodeInfoMap[node.Node().Name]; found {
+		data.removeNode(node.Node().Name)
+	}
+	if _, found := data.modifiedNodeInfoMap[node.Node().Name]; found {
+		data.removeNode(node.Node().Name)
+	}
+
+	return data.addNodeInfo(node)
+}
+
+func (data *internalDeltaSnapshotData) removeNode(nodeName string) error {
+	_, foundInDelta := data.addedNodeInfoMap[nodeName]
+	if foundInDelta {
+		// If node was added within this delta, delete this change.
+		delete(data.addedNodeInfoMap, nodeName)
+	}
+
+	if _, modified := data.modifiedNodeInfoMap[nodeName]; modified {
+		// If node was modified within this delta, delete this change.
+		delete(data.modifiedNodeInfoMap, nodeName)
+	}
+
+	_, foundInBase := data.baseData.getNodeInfo(nodeName)
+	if foundInBase {
+		// If node was found in the underlying data, mark it as deleted in delta.
+		data.deletedNodeInfos[nodeName] = true
+	}
+
+	if !foundInBase && !foundInDelta {
+		// Node not found in the chain.
+		return errNodeNotFound
+	}
+
+	// Maybe consider deleting from the lists instead. Maybe not.
+	data.clearCaches()
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) addPod(pod *apiv1.Pod, nodeName string) error {
+	dni, found := data.getNodeInfoLocal(nodeName)
+	if !found {
+		bni, found := data.baseData.getNodeInfo(nodeName)
+		if !found {
+			return errNodeNotFound
+		}
+		dni = bni.Clone()
+		data.modifiedNodeInfoMap[nodeName] = dni
+		data.clearCaches()
+	}
+
+	dni.AddPod(pod)
+	if data.podList != nil || data.havePodsWithAffinity != nil {
```

Comment:
Why this if? Calling clearPodCaches if this is if not true does nothing and I don't believe it carries any cost.